src/DotLiquid/Util/StrFTime.cs

```diff
@@ -18,85 +20,197 @@ public static class StrFTime
+                x => StrFTimeMatchEvaluator(
+                    x.Groups[0].Value,
+                    x.Groups["flag"].Captures.Cast<Capture>().Select(y => y.Value).ToList(),
+                    x.Groups["width"].Captures.Cast<Capture>().Select(y => (int?)Convert.ToInt32(y.Value)).FirstOrDefault(),
+                    x.Groups["directive"].Captures.Cast<Capture>().Select(y => y.Value).FirstOrDefault(),
+                    dateTime
+                    ));
         }
 
         /// <summary>
         /// Formats a date using a ruby date format string
         /// </summary>
-        /// <param name="dateTime"></param>
+        /// <param name="dateTimeOffset"></param>
         /// <param name="pattern"></param>
         /// <returns></returns>
-        public static string ToStrFTime(this DateTimeOffset dateTime, string pattern)
+        public static string ToStrFTime(this DateTimeOffset dateTimeOffset, string pattern)
+        {
+            return Regex.Replace(pattern, @"%(?<flag>[-_0^:#])*(?<width>[1-9][0-9]*)?(?<directive>[a-zA-Z%\+])",
+                x => StrFTimeMatchEvaluator(
+                    x.Groups[0].Value,
+                    x.Groups["flag"].Captures.Cast<Capture>().Select(y => y.Value).ToList(),
+                    x.Groups["width"].Captures.Cast<Capture>().Select(y => (int?)Convert.ToInt32(y.Value)).FirstOrDefault(),
+                    x.Groups["directive"].Captures.Cast<Capture>().Select(y => y.Value).FirstOrDefault(),
+                    dateTimeOffset
+                    ));
+        }
+
+        private static String StrFTimeMatchEvaluator(String orig, IEnumerable<String> flags, int? width, String directive, object obj)
+        {
+            var result = orig;
+            directive = PreProcessDirective(directive, flags, width);
+
+            bool isDateTimeOffset = obj is DateTimeOffset;
+            if (OffsetFormats.ContainsKey(directive) && isDateTimeOffset)
+                result = OffsetFormats[directive].Invoke((DateTimeOffset)obj);
+            else if (Formats.ContainsKey(directive))
+                result = Formats[directive].Invoke(isDateTimeOffset ? ((DateTimeOffset)obj).DateTime : (DateTime)obj);
+            else
+                return orig; // This is an unspecified directive
+
+            return flags.ToList().Aggregate(result, (current, flag) => ApplyFlag(flag, (width ?? 2), current));
+        }
+
+        // Pre-process the directive for some of the quirkier cases, such as '%:z' and '%3N'.
+        private static String PreProcessDirective(String directive, IEnumerable<String> flags, int? width)
         {
-            var output = new System.Text.StringBuilder();
+            String result = directive;
+            String flagString = String.Join("", flags);
+            if ("z".Equals(directive) && ":".Equals(flagString))
+            {
+                result = flagString + directive;

```

Comment:
it may be a cleaner pattern to return directly the result rather than mutate the 'result' local variable.

Reply:
Replaced local variable mutation with return statements.